x-pack/auditbeat/module/system/packages/packages.go

```diff
@@ -0,0 +1,356 @@
+)
+
+func init() {
+	mb.Registry.MustAddMetricSet(moduleName, metricsetName, New,
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet collects data about the host.
+type MetricSet struct {
+	mb.BaseMetricSet
+	config config.Config
+	cache  *cache.Cache
+	log    *logp.Logger
+}
+
+// Package represents information for a package.
+type Package struct {
+	Name        string
+	Version     string
+	Release     string
+	Arch        string
+	License     string
+	InstallTime time.Time
+	Size        uint64
+	Summary     string
+	URL         string
+}
+
+// Hash creates a hash for Package.
+func (pkg Package) Hash() string {
+	// Could use real hash e.g. FNV if there is an advantage
+	return pkg.Name + pkg.InstallTime.String()
+}
+
+func (pkg Package) toMapStr() common.MapStr {
+	return common.MapStr{
+		"package.name":        pkg.Name,
+		"package.version":     pkg.Version,
+		"package.release":     pkg.Release,
+		"package.arch":        pkg.Arch,
+		"package.license":     pkg.License,
+		"package.installtime": pkg.InstallTime,
+		"package.size":        pkg.Size,
+		"package.summary":     pkg.Summary,
+		"package.url":         pkg.URL,
+	}
+}
+
+// New constructs a new MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The %v/%v dataset is experimental", moduleName, metricsetName)
+
+	config := config.NewDefaultConfig()
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrapf(err, "failed to unpack the %v/%v config", moduleName, metricsetName)
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+		log:           logp.NewLogger(moduleName),
+	}
+
+	if config.ReportChanges {
+		ms.cache = cache.New()
+	}
+
+	return ms, nil
+}
+
+// Fetch collects data about the host. It is invoked periodically.
+func (ms *MetricSet) Fetch(report mb.ReporterV2) {
+	packages, err := getPackages()
+	if err != nil {
+		ms.log.Error(err)
+		report.Error(err)
+	}
+	if packages == nil {
+		return
+	}
+
+	if ms.cache != nil && !ms.cache.IsEmpty() {
+		installed, removed := ms.cache.DiffAndUpdateCache(packages)
+
+		for _, pkgInfo := range installed {
+			report.Event(mb.Event{
+				MetricSetFields: common.MapStr{
+					"status":   "installed",
+					"packages": pkgInfo.(Package).toMapStr(),
+				},
+			})
+		}
+
+		for _, pkgInfo := range removed {
+			report.Event(mb.Event{
+				MetricSetFields: common.MapStr{
+					"status":   "removed",
+					"packages": pkgInfo.(Package).toMapStr(),
+				},
+			})
+		}
+	} else {
+		// Report all installed packages
+		var pkgInfos []common.MapStr
+
+		for _, pkgInfo := range packages {
+			pkgInfos = append(pkgInfos, pkgInfo.(Package).toMapStr())
+		}
+
+		report.Event(mb.Event{
+			MetricSetFields: common.MapStr{
+				"packages": pkgInfos,
+			},
+		})
+
+		if ms.cache != nil {
+			// This will initialize the cache with the current packages
+			ms.cache.DiffAndUpdateCache(packages)
+		}
+	}
+}
+
+func getPackages() ([]cache.Cacheable, error) {
```

Comment:
I tried that at first, but I believe it's not possible to assign a slice of a struct to a slice of an interface - even if the struct is implementing the interface (see [this answer on SO](https://stackoverflow.com/a/12994852) and [this article on the Golang wiki)](https://github.com/golang/go/wiki/InterfaceSlice)

I guess we could return `[]Package` and explicitly convert it to `[]cache.Cacheable` (`make` a new slice and iterate-copy over the packages). Though that doesn't seem very nice either.

What do you think?

Reply:
Correct, you cannot convert the type of a slice.

It will be better to retain the `Package` type especially since the code needs the type information (there are several type assertions going back from `Cacheable` -> `Package` and each on incurs a cpu cost (but not memory)).

Converting `[]Package` to `[]Cacheable` won't be that great since it cause a fair bit of copying. Going from `[]*Package` to `[]Cacheable` would faster since it only needs to copy the  uintptr value (8 bytes). You'd probably want to define the `Hash()` method on a pointer receiver.

You can see from these benchmarks that converting from `[]*Package` to `[]Cacheable` is cheap if done with some consideration of the data types and allocations.

```
BenchmarkAllocateInterfaceSlice-8         200000              8674 ns/op           16192 B/op        101 allocs/op
BenchmarkAllocateInterfaceSliceWithPointers-8            3000000               420 ns/op            1792 B/op          1 allocs/op
BenchmarkAllocateInterfaceSliceOnceWithPointers-8       20000000               110 ns/op               0 B/op          0 allocs/op


func BenchmarkAllocateInterfaceSlice(b *testing.B) {
	pkg := make([]Package, 100)
	var c []Cacheable

	for i := 0; i < b.N; i++ {
		c = make([]Cacheable, len(pkg))
		for i, p := range pkg {
			c[i] = p
		}
	}
}

func BenchmarkAllocateInterfaceSliceWithPointers(b *testing.B) {
	pkg := make([]*Package, 100)
	var c []Cacheable

	for i := 0; i < b.N; i++ {
		c = make([]Cacheable, len(pkg))
		for i, p := range pkg {
			c[i] = p
		}
	}
}

func BenchmarkAllocateInterfaceSliceOnceWithPointers(b *testing.B) {
	var c []Cacheable // Reuse this across all Fetches.
	pkg := make([]*Package, 100)

	for i := 0; i < b.N; i++ {
		// Resize the slice to len=0.
		c = c[:0]
		// See if the slice needs to be re-allocated based on number of pkgs.
		if cap(c) < len(pkg) {
			c = make([]Cacheable, 0, len(pkg))
		}

		for _, p := range pkg {
			c = append(c, p)
		}
	}
}
```

Reply:
Thanks for running the benchmark. I've added `convertToCacheable` methods before calling `cache.DiffAndUpdateCache`.